godot-core/src/task/async_runtime.rs

```diff
@@ -14,9 +14,251 @@ use std::sync::Arc;
+        runtime_storage: None,
+        registration_attempted: false,
+    }) };
+}
+
+/// Register an async runtime integration with gdext for the current thread
+///
+/// This must be called before using any async functions like `#[async_func]` on this thread.
+/// Each thread can have its own runtime registration.
+///
+/// # Errors
+///
+/// Returns an error if a runtime has already been registered for this thread.
+///
+/// # Example
+///
+/// ```rust
+/// use your_runtime_integration::YourRuntimeIntegration;
+///
+/// // Register your runtime at application startup
+/// gdext::task::register_runtime::<YourRuntimeIntegration>()?;
+///
+/// // Now async functions will work on this thread
+/// ```
+pub fn register_runtime<T: AsyncRuntimeIntegration>() -> Result<(), String> {
+    RUNTIME_REGISTRY.with(|registry| {
+        let mut registry = registry.borrow_mut();
+
+        if registry.registration_attempted {
+            return Err("Async runtime has already been registered for this thread".to_string());
+        }
+
+        registry.registration_attempted = true;
+
+        // Create the runtime immediately during registration
+        let (runtime_instance, handle) = T::create_runtime()?;
+
+        // Clone the handle for the closure
+        let handle_clone = handle.clone();
+
+        // Create the storage structure with context management
+        let storage = RuntimeStorage {
+            _runtime_instance: runtime_instance,
+            with_context: Box::new(move |f| T::with_context(&handle_clone, f)),
+        };
+
+        registry.runtime_storage = Some(storage);
+        Ok(())
+    })
+}
+
+/// Check if a runtime is registered for the current thread
+pub fn is_runtime_registered() -> bool {
+    RUNTIME_REGISTRY.with(|registry| registry.borrow().runtime_storage.is_some())
+}
+
+// ----------------------------------------------------------------------------------------------------------------------------------------------
+
+// Enhanced Error Handling
+
+/// Errors that can occur during async runtime operations
+#[derive(Debug, Clone)]
+pub enum AsyncRuntimeError {
+    /// Runtime has been deinitialized (during engine shutdown)
+    RuntimeDeinitialized,
+    /// Task was canceled while being polled
+    TaskCanceled { task_id: u64 },
+    /// Task panicked during polling
+    TaskPanicked { task_id: u64, message: String },
+    /// Task slot is in an invalid state
+    InvalidTaskState {
+        task_id: u64,
+        expected_state: String,
+    },
+    /// No async runtime has been registered
+    NoRuntimeRegistered,
+    /// Task spawning failed
+    TaskSpawningFailed { reason: String },
+    /// Signal emission failed
+    SignalEmissionFailed { task_id: u64, reason: String },
+    /// Thread safety violation
+    ThreadSafetyViolation {
+        expected_thread: ThreadId,
+        actual_thread: ThreadId,
+    },
+}
```

Comment:
This is a very detailed error API. I would probably keep this private in the beginning -- and wrapped by a public-facing struct:
```rs
pub struct AsyncRuntimeError {
     inner: AsyncInnerError, // your enum
}
```
Then all the traits (`Display` etc.) can appear on the public-facing type. This reduces API complexity quite a bit. If people _need_ to differentiate those programmatically, we can always gradually expose the details, but then we start simple.